tokio-util/src/io/inspect.rs

```diff
@@ -0,0 +1,115 @@
+use futures_core::ready;
+use pin_project_lite::pin_project;
+use std::{
+    io::{IoSlice, Result},
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::io::{AsyncRead, AsyncWrite, ReadBuf};
+
+pin_project! {
+    /// An adapter that lets you inspect the data that's being read.
+    ///
+    /// This is useful for things like hashing data as it's read in.
+    pub struct InspectReader<R: AsyncRead, F: FnMut(&[u8])> {
+        #[pin]
+        reader: R,
+        f: F,
+    }
+}
+
+impl<R: AsyncRead, F: FnMut(&[u8])> InspectReader<R, F> {
+    /// Create a new InspectReader, wrapping `reader` and calling `f` for the
+    /// new data supplied by each read call.
+    pub fn new(reader: R, f: F) -> InspectReader<R, F> {
+        InspectReader { reader, f }
+    }
```

Comment:
It's more convenient for the user to not require the traits for `into_inner`. Arguably the same applies to `new`, but not requiring it for `new` gives worse error messages.
```suggestion
impl<R, F> InspectReader<R, F> {
    /// Create a new InspectReader, wrapping `reader` and calling `f` for the
    /// new data supplied by each read call.
    pub fn new(reader: R, f: F) -> InspectReader<R, F> 
    where
        R: AsyncRead,
        F: FnMut(&[u8]),
    {
        InspectReader { reader, f }
    }
```